scripts/modules_used.lean

```diff
@@ -0,0 +1,101 @@
+import all
+import system.io
+import tactic.algebra
+
+/-!
+# Find all imported modules which are used by the declarations in the target module.
+
+```
+lean --run scripts/modules_used.lean data.nat.order.basic
+```
+
+returns
+
+```
+order.synonym
+order.rel_classes
+order.monotone
+order.lattice
+order.heyting.basic
+order.bounded_order
+order.boolean_algebra
+order.basic
+logic.nontrivial
+logic.nonempty
+logic.is_empty
+logic.function.basic
+logic.equiv.defs
+logic.basic
+data.subtype
+data.set.basic
+data.nat.order.basic
+data.nat.cast.defs
+data.nat.basic
+algebra.ring.defs
+algebra.order.zero_le_one
+algebra.order.sub.defs
+algebra.order.sub.canonical
+algebra.order.ring.lemmas
+algebra.order.ring.defs
+algebra.order.ring.canonical
+algebra.order.monoid.lemmas
+algebra.order.monoid.defs
+algebra.order.monoid.canonical.defs
+algebra.order.monoid.cancel.defs
+algebra.group_with_zero.defs
+algebra.group.defs
+algebra.group.basic
+algebra.covariant_and_contravariant
+```
+
+This is useful for finding imports which might be removable.
+-/
+
+open tactic declaration environment io io.fs
+
+meta def tactic.get_decls_used (env : environment) : name → name_set → tactic name_set
+| n ns := if ns.contains n then pure ns else do
+  d ← env.get n,
+  -- Add `n` to the accumulated name set.
+  let ns := ns.insert n,
+  -- Run `get_decls_used` on any ancestors of `n` (if `n` is a structure)
+  ancestors ← get_ancestors n,
+  ns ← ancestors.mfoldl (λ ns n, tactic.get_decls_used n ns) ns,
+  -- Now traverse the body of the declaration, processing any constants.
+  let process (v : expr) : tactic (name_set) :=
+    v.fold (pure ns) $ λ e _ r, r >>= λ ns,
+      if e.is_constant then tactic.get_decls_used e.const_name ns else pure ns,
+  match d with
+  | (declaration.defn _ _ _ v _ _) := process v
+  | (declaration.thm _ _ _ v)      := process v.get
+  | _ := pure ns
+  end
+
+meta def tactic.get_modules_used_by_theorems_in (tgt : string) : tactic (list string) :=
+do env ← tactic.get_env,
+  ns ← env.fold (pure mk_name_set) (λ d r,
+    if env.decl_olean d.to_name = some tgt then
+      r >>= tactic.get_decls_used env d.to_name
+    else r),
+  let mods := ns.fold native.mk_rb_set (λ n mods,
+    match env.decl_olean n with
+    | some mod := mods.insert mod
+    | none := mods
+    end),
+  pure mods.to_list
```

Comment:
Does it make sense to sort alphabetically?

Reply:
I'm using this to highlight nodes in a graph, so the order becomes immaterial.

Reply:
Looks like it's actually already sorted anyway, just in reverse order. Maybe add `.reverse` for human use?

Reply:
Done

Reply:
Can you update the example in the PR description and/or docstring?